src/mod_disco.erl

```diff
@@ -197,57 +193,74 @@ process_sm_iq_info(Acc, From, To, #iq{type = get, lang = Lang, sub_el = SubEl} =
+-spec disco_sm_identity(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: mongoose_disco:identity_acc(),
+    Params :: map(),
+    Extra :: map().
+disco_sm_identity(Acc = #{to_jid := JID}, _, _) ->
+    NewAcc = case ejabberd_auth:does_user_exist(JID) of
         true -> mongoose_disco:add_identities([#{category => <<"account">>,
                                                  type => <<"registered">>}], Acc);
         false -> Acc
-    end.
-
--spec disco_local_items(mongoose_disco:item_acc()) -> mongoose_disco:item_acc().
-disco_local_items(Acc = #{host_type := HostType, from_jid := From, to_jid := To, node := <<>>}) ->
+    end,
+    {ok, NewAcc}.
+
+-spec disco_local_items(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: mongoose_disco:item_acc(),
+    Params :: map(),
+    Extra :: map().
+disco_local_items(Acc = #{host_type := HostType, from_jid := From, to_jid := To, node := <<>>}, _, _) ->
     ReturnHidden = should_return_hidden(HostType, From),
     Subdomains = get_subdomains(To#jid.lserver),
     Components = get_external_components(To#jid.lserver, ReturnHidden),
     ExtraDomains = get_extra_domains(HostType),
     Domains = Subdomains ++ Components ++ ExtraDomains,
-    mongoose_disco:add_items([#{jid => Domain} || Domain <- Domains], Acc);
-disco_local_items(Acc) ->
-    Acc.
+    {ok, mongoose_disco:add_items([#{jid => Domain} || Domain <- Domains], Acc)};
+disco_local_items(Acc, _, _) ->
+    {ok, Acc}.
 
--spec disco_sm_items(mongoose_disco:item_acc()) -> mongoose_disco:item_acc().
-disco_sm_items(Acc = #{to_jid := To, node := <<>>}) ->
+-spec disco_sm_items(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: mongoose_disco:item_acc(),
+    Params :: map(),
+    Extra :: map().
+disco_sm_items(Acc = #{to_jid := To, node := <<>>}, _, _) ->
     Items = get_user_resources(To),
-    mongoose_disco:add_items(Items, Acc);
-disco_sm_items(Acc) ->
-    Acc.
+    {ok, mongoose_disco:add_items(Items, Acc)};
+disco_sm_items(Acc, _, _) ->
+    {ok, Acc}.
 
--spec disco_local_features(mongoose_disco:feature_acc(),
-                           map(),
-                           map()) -> {ok, mongoose_disco:feature_acc()}.
+-spec disco_local_features(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: mongoose_disco:feature_acc(),
+    Params :: map(),
+    Extra :: map().
 disco_local_features(Acc = #{node := <<>>}, _, _) ->
     {ok, mongoose_disco:add_features([<<"iq">>, <<"presence">>, <<"presence-invisible">>], Acc)};
 disco_local_features(Acc, _, _) ->
     {ok, Acc}.
 
 %% @doc Support for: XEP-0157 Contact Addresses for XMPP Services
--spec disco_info(mongoose_disco:info_acc()) -> mongoose_disco:info_acc().
-disco_info(Acc = #{host_type := HostType, module := Module, node := <<>>}) ->
+-spec disco_info(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: mongoose_disco:info_acc(),
+    Params :: map(),
+    Extra :: map().
+disco_info(Acc = #{host_type := HostType, module := Module, node := <<>>}, _, _) ->
```

Comment:
Hmmm... all these handlers expose the host_type twice, once here and once in the extra param, seems like redundancy to correct for the future 🤔